relay-redis/src/real.rs

```diff
@@ -34,28 +35,68 @@ enum ConnectionInner<'a> {
     Single(&'a mut redis::Connection),
 }
 
-impl ConnectionLike for ConnectionInner<'_> {
-    fn req_packed_command(&mut self, cmd: &[u8]) -> redis::RedisResult<redis::Value> {
+impl ConnectionInner<'_> {
+    fn get_connections(&mut self) -> Vec<&mut (dyn ConnectionLike + Send)> {
         match self {
-            ConnectionInner::Cluster(ref mut con) => con.req_packed_command(cmd),
+            ConnectionInner::Cluster(con) => vec![con as &mut (dyn ConnectionLike + Send)],
             ConnectionInner::MultiWrite {
                 primary: primary_connection,
                 secondaries: secondary_connections,
             } => {
-                let primary_result = primary_connection.req_packed_command(cmd);
+                let mut connections = primary_connection.get_connections();
                 for secondary_connection in secondary_connections.iter_mut() {
-                    if let Err(error) = secondary_connection.req_packed_command(cmd) {
+                    connections.extend(secondary_connection.get_connections());
+                }
+
+                connections
+            }
+            ConnectionInner::Single(con) => vec![con as &mut (dyn ConnectionLike + Send)],
+        }
+    }
+}
+
+impl ConnectionLike for ConnectionInner<'_> {
+    fn req_packed_command(&mut self, cmd: &[u8]) -> redis::RedisResult<redis::Value> {
+        let mut results = thread::scope(|s| {
+            let handles: Vec<_> = self
+                .get_connections()
+                .into_iter()
+                .map(|connection| s.spawn(move || connection.req_packed_command(cmd)))
+                .collect();
+
+            handles
+                .into_iter()
+                .map(|handle| handle.join())
+                .collect::<Vec<_>>()
+        })
+        .into_iter();
+
+        let Some(Ok(primary_result)) = results.next() else {
+            relay_log::error!("sending cmd to primary Redis instance failed");
+            return Err(io::Error::new(
+                ErrorKind::Other,
+                "failed to send cmd to primary Redis instance",
+            )
+            .into());
+        };
```

Comment:
Is this necessary? The result is an error anyways and we can let the caller deal with the handle, we don't need an extra layer of logging and mapping of errors here.

Reply:
It was handling the case in which the thread making the connection had an error.

Reply:
I would exclude the possibility of `None` here by changing the return type of `get_connections` to `(Whatever, Vec<Whatever>)`. That also prevents allocations in the more common case where there's only a primary. And then do all the logging in the spawned tasks as David suggested (no explicit joining on secondaries).

Reply:
Yes, I was doing this. Similar to how I designed the other types.